add a promise return to the programmatic modal open.

urkle opened the pull request:
<!-- Thank you for helping Oruga! -->

Proposal for #225

Depends on PR #315 

## Proposed Changes

- Modal instance now has a "promise" property that when using $oruga.modal.open contains a promise that will resolve to the value passed to the $emit('close', value)
- when a modal is canceled, the promise resolves {action: 'cancel', method: 'x'}

jtommy commented:
The idea is good but it's a breaking change.
I developed a similar feature in Buefy dialog in order to support Promise without introducing a breaking change, try to take a look :)

jtommy commented:
It's always a breaking change because Promise is not supported by all browsers (Vue 2 supports IE11).
We might think to introduce it in the next breaking only for next version, I think Buefy dialog implementation is a good compromesso, don't you?

urkle commented:
> 

Since it only extends the Modal (e.g. it doesn't change the return contract) we can do one of the following
1) have an option to turn it on an off (e.g. whether I set the promise property or not)
   - but I don't like changing the return contract based on a global variable lik the Buefy implementation does as that means you either can have a promise OR gain access to the instance but never both.
2) if there is no "Promise" available then we don't set it.

IMHO option 2 sounds the best.  As in reality anyone who  is using "promises" in their code (or async/await) and wanting to support IE 11 will have a polyfill in their code builds so things will just work.

I can do a quick change for that.

urkle commented:
@jtommy I added a guard around the promise usage for the Vue 2 implementation (not the Vue 3).

Further I added extra comments in the example to clarify that promises can't be used on IE 11 unless a polyfill is added to your build.

I also added a comment in the example to show that "parent: this" is only needed for Vue 2 and should not be included in Vue 3.

jtommy commented:
@urkle I agree with you.
I am not sure about the change of programmatic prop but it works.
After that PR, are you able to implement a dialog? For example a prompt?

urkle commented:
>  I am not sure about the change of programmatic prop but it works.
Since it wasn't even documented, I figured it was a simple way to pass those callbacks into the dialog easily. If you prefer I can add it as a different prop name, but utilizing the programmatic was simple (though I need to update the property definittion for the IE11 tweak)

> After that PR, are you able to implement a dialog? For example a prompt?

Yes, with this change I can implement my dialog prompts the way I want.  Actually I'm going to "amend" this PR once more before you merge it to separate out the examples so that there is a separate "programmatic" example using promises from the one that is basic. And I'll have that later one implement a simple "prompt" dialog.

Now, a followup question,
Can we apply this "pattern" to notifications and loading? (so developers can essentially tab into a async/await approach to handling when it is closed/dimissed).

urkle commented:
@jtommy all pushed up.

Changes I made since the last one are
1) adjusted "type" of promise property so that Vue doesn't yell about it (it's undefined meaning any for now. we *could* use Promise in the vue3 one if we wanted)
2) I split the programmatic promise/async/await example to a separate example grouping and restored the previous one. The new example shows a basic implementation of a "prompt" (yes/no) style dialog.  It will "return" either action: 'yes', action: 'no', or action: 'cancel' (if they hit off the dialog or hit the X icon).
